File: iphone/msgs/Classes/MessageStore.m
//
//  MessageStore.m
//  msgs
//
//  Created by Lewis Zimmerman on 11-01-15.
//  Copyright 2011 Qorporation. All rights reserved.
//

#import "MessageStore.h"
#import "Atlas.h"
#import "StringUtil.h"

#import <Three20Core/TTCorePreprocessorMacros.h>

static MessageStore* instance = nil;

///////////////////////////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////////////////////////
@implementation MessageStore

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)dealloc {
	TT_RELEASE_SAFELY(fmdb);
	
	[super dealloc];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)setupDB {	
	NSString* accountDBPath = [StringUtil filePathInDocumentsDirectoryForFileName:kMessageStoreDB];
	fmdb = [[FMDatabase alloc] initWithPath:accountDBPath];
	if(![fmdb open]) {
		NSLog(@"Could not open the DB for whatever reason about to remove and copy the base file...here was the error:");
		NSLog(@"DB Err %d: %@", [fmdb lastErrorCode], [fmdb lastErrorMessage]);
	}
	
	[fmdb setLogsErrors:YES];
	[fmdb executeUpdate:@"PRAGMA CACHE_SIZE=1000"];
	[fmdb executeUpdate:@"PRAGMA encoding = \"UTF-8\""];
	[fmdb executeUpdate:@"PRAGMA auto_vacuum=1"];
	[fmdb executeUpdate:@"PRAGMA synchronous=NORMAL"];
	
	[fmdb executeUpdate:@"CREATE TABLE IF NOT EXISTS message (pk INTEGER PRIMARY KEY, "
						 "id INTEGER, "
						 "sender INTEGER, "
						 "conversation INTEGER, "
						 "time INTEGER, "
						 "body VARCHAR(512))"];
	
	[fmdb executeUpdate:@"CREATE UNIQUE INDEX IF NOT EXISTS message_id ON message(id);"];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (NSDictionary*)resultToDict:(FMResultSet*)rs {
	return [[NSDictionary alloc] initWithObjectsAndKeys:
			[NSNumber numberWithInt:[rs intForColumn:@"id"]], @"id",
			[NSNumber numberWithInt:[rs intForColumn:@"sender"]], @"sender",
			[NSNumber numberWithInt:[rs intForColumn:@"conversation"]], @"conversation",
			[NSNumber numberWithInt:[rs intForColumn:@"time"]], @"time",
			[rs stringForColumn:@"body"], @"body",
			nil];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (id)init {
	if (self = [super init]) {	
		[self setupDB];
	}
	
	return self;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
+ (id)get {
	@synchronized(self) {
		if (instance == nil) {
			instance = [[self alloc] init];
		}
	}
	
	return instance;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (void)setDelegate:(id<MessageStoreDelegate>)storeDelegate {
	delegate = storeDelegate;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (bool)setMessage:(NSNumber*)message sender:(NSNumber*)sender conversation:(NSNumber*)conversation time:(NSNumber*)time body:(NSString*)body {
	bool res = [fmdb executeUpdate:@"INSERT OR REPLACE INTO message(id, sender, conversation, time, body) VALUES(?, ?, ?, ?, ?)", 
				message, sender, conversation, time, body];
	
	if (delegate != nil && [[delegate listeningConversation] isEqual:conversation]) {
		[delegate receivedMessage:[self getByID:message]];
	}
	
	return res;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (bool)setMessage:(NSDictionary*)payload {
	return [self setMessage:[NSNumber numberWithLongLong:[[payload objectForKey:@"message"] longLongValue]] 
					 sender:[NSNumber numberWithLongLong:[[payload objectForKey:@"sender"] longLongValue]] 
			   conversation:[NSNumber numberWithLongLong:[[payload objectForKey:@"conversation"] longLongValue]] 
					   time:[NSNumber numberWithLongLong:[[payload objectForKey:@"time"] longLongValue]]
					   body:[payload objectForKey:@"body"]];
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (NSDictionary*)getByID:(NSNumber*)ident {
	FMResultSet* rs = [fmdb executeQuery:@"SELECT * FROM message WHERE id = ?", ident];
	if ([rs next]) {
		NSDictionary* message = [self resultToDict:rs];
		
		[rs close];	
		return message;
	}
	
	[rs close];	
	return nil;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (NSDictionary*)getMostRecentMessageForConversation:(NSNumber*)ident {
	FMResultSet* rs = [fmdb executeQuery:@"SELECT * FROM message WHERE conversation = ?", ident];
	if ([rs next]) {
		NSDictionary* message = [self resultToDict:rs];
		
		[rs close];	
		return message;
	}
	
	[rs close];	
	return nil;
}

///////////////////////////////////////////////////////////////////////////////////////////////////
- (NSArray*)messagesForConversation:(NSNumber*)ident {
	FMResultSet* rs = [fmdb executeQuery:@"SELECT * FROM message WHERE conversation = ?", ident];
	NSMutableArray* list = [[NSMutableArray alloc] init];
	
	while ([rs next]) {
		[list addObject:[self resultToDict:rs]];
	}
	
	[rs close];	
	return list;
}

@end

File: android/client/src/com/qorporation/msgs/client/android/store/MessageStore.java
package com.qorporation.msgs.client.android.store;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import org.json.JSONObject;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteStatement;

import com.qorporation.msgs.client.android.store.delegates.MessageStoreDelegate;

public class MessageStore extends Store {
	private static MessageStore instance = null;
	public static synchronized MessageStore getInstance() {
		if (instance == null) instance = new MessageStore(Store.DEFAULT_CONTEXT);
		return instance;
	}
	
	private MessageStoreDelegate delegate = null;
	public void setDelegate(MessageStoreDelegate delegate) { this.delegate = delegate; }
	
	private MessageStore(Context context) {
		super(context, "message.db", Arrays.asList(
				"CREATE TABLE IF NOT EXISTS message (pk INTEGER PRIMARY KEY, "
				+ "id INTEGER, "
				+ "sender INTEGER, "
				+ "conversation INTEGER, "
				+ "time INTEGER, "
				+ "body VARCHAR(512))",

				"CREATE UNIQUE INDEX IF NOT EXISTS message_id ON message(id);"
		));
	}

	private HashMap<String, Object> rowToItem(Cursor cursor) {
		HashMap<String, Object> val = new HashMap<String, Object>();
		
		try {
			val.put("id", new Integer(cursor.getInt(1)));
			val.put("sender", new Integer(cursor.getInt(2)));
			val.put("conversation", new Integer(cursor.getInt(3)));
			val.put("time", new Long(cursor.getLong(4)));
			val.put("body", cursor.getString(5));
		} catch (Exception e) {
		}
		
		return val;
	}
	
	public boolean setMessage(int id, int sender, int conversation, long time, String body) {
		boolean ret = false;
		
		try {
			SQLiteStatement s = this.db.compileStatement("INSERT OR REPLACE INTO message(id, sender, conversation, time, body) VALUES(?, ?, ?, ?, ?)");
            s.bindLong(1, id);
            s.bindLong(2, sender);
            s.bindLong(3, conversation);
            s.bindLong(4, time);
            s.bindString(5, body);
            
            ret = s.executeInsert() > 0;
            s.close();
            
            if (this.delegate != null) {
            	if (this.delegate.listeningConversation() == -1 
            		|| this.delegate.listeningConversation() == conversation) {
            		this.delegate.onMessageEvent(this.getByID(id));
            	}
            }
		} catch (Exception e) {
		}
		
		return ret;
	}
	
	public boolean setMessage(JSONObject val) throws Exception {
		return setMessage(val.getInt("message"), val.getInt("sender"), val.getInt("conversation"), val.getLong("time"), val.getString("body"));
	}
    
    public HashMap<String, Object> getByID(int id) {
        try {          
            Cursor c = this.db.query("message", new String[] { "id", "sender", "conversation", "time", "body" }, "id = ?", new String[] { Integer.toString(id) }, null, null, null);

            if (c.moveToNext()) {                    
            	this.rowToItem(c);                                
            }

            c.close();
        } catch(Exception e) {
        } 
        
    	return null;
    }
    
    public HashMap<String, Object> getMostRecentForConversation(int conversation) {
        try {          
            Cursor c = this.db.query("message", new String[] { "id", "sender", "conversation", "time", "body" }, "conversation = ?", new String[] { Integer.toString(conversation) }, null, null, "time DESC");

            if (c.moveToNext()) {                    
            	this.rowToItem(c);                                
            }

            c.close();
        } catch(Exception e) {
        } 
        
    	return null;
    }
    
    public List<HashMap<String, Object>> messagesForConversation(int conversation) {
        ArrayList<HashMap<String, Object>> list = new ArrayList<HashMap<String, Object>>();
        
        try {          
            Cursor c = this.db.query("message", new String[] { "id", "sender", "conversation", "time", "body" }, "conversation = ?", new String[] { Integer.toString(conversation) }, null, null, "time DESC");

            while (c.moveToNext()) {                    
                list.add(this.rowToItem(c));                                  
            }

            c.close();
        } catch(Exception e) {
        } 
        
        return list;          
    }
    
    public List<HashMap<String, Object>> messagesForConversationAfter(int conversation, long time) {
        ArrayList<HashMap<String, Object>> list = new ArrayList<HashMap<String, Object>>();
        
        try {          
            Cursor c = this.db.query("message", new String[] { "id", "sender", "conversation", "time", "body" }, "conversation = ? AND time > ?", new String[] { Integer.toString(conversation), Long.toString(time) }, null, null, "time DESC");

            while (c.moveToNext()) {                    
                list.add(this.rowToItem(c));                                  
            }

            c.close();
        } catch(Exception e) {
        } 
        
        return list;          
    }

}
